File: frontend/src/engine/ifc/ifcExtractor.ts
/** Port of backend/app/extraction/ifc_extractor.py. */

import * as WebIFC from "web-ifc";
import type { IfcAPI } from "web-ifc";
import type { AxisReference } from "../axisReference";
import type { Band, ElementType, Side, SourceMethod, StateKind, WidthSample } from "../../types/domain";
import { attrRef, attrRefList, attrString } from "./webIfcClient";
import { pavementWidthSamples, type PlanWidthSample } from "./ifcGeometry";
import { maxOf, minOf, pushAll } from "../arrayUtils";

const KEYWORD_HINTS: [RegExp, ElementType][] = [
  [/bau/i, "bau"],
  [/accot/i, "accotement"],
  [/tr.{0,4}oir/i, "trottoir"],
  [/cycl/i, "cycle"],
  [/voie|chauss/i, "voie"],
  [/tpc|median|terre.?plein/i, "tpc"],
];

export function guessElementType(typeName: string): [ElementType, number] {
  for (const [pattern, elementType] of KEYWORD_HINTS) {
    if (pattern.test(typeName)) return [elementType, 0.3];
  }
  return ["non_utilise", 0.1];
}

interface PavementGroup {
  typeName: string;
  expressIds: number[];
}

function pavementTypeName(api: IfcAPI, modelID: number, line: Record<string, unknown>, expressID: number): string {
  const relTypeIds = api.GetLineIDsWithType(modelID, WebIFC.IFCRELDEFINESBYTYPE, true);
  for (let i = 0; i < relTypeIds.size(); i++) {
    const rel = api.GetLine(modelID, relTypeIds.get(i)) as Record<string, unknown>;
    const related = attrRefList(rel, "RelatedObjects");
    if (related.includes(expressID)) {
      const typeId = attrRef(rel, "RelatingType");
      if (typeId !== null) {
        const typeLine = api.GetLine(modelID, typeId) as Record<string, unknown>;
        return attrString(typeLine, "Name") ?? `Type-${typeId}`;
      }
    }
  }
  return attrString(line, "Name") ?? `Pavement-${expressID}`;
}

function roadAncestorName(api: IfcAPI, modelID: number, expressID: number, parentMap: Map<number, number>): string | null {
  let current = expressID;
  for (let depth = 0; depth < 6; depth++) {
    const parent = parentMap.get(current);
    if (parent === undefined) return null;
    const parentLine = api.GetLine(modelID, parent) as Record<string, unknown>;
    if (parentLine.type === WebIFC.IFCROAD) return attrString(parentLine, "Name");
    current = parent;
  }
  return null;
}

function matchesState(name: string | null, state: StateKind): boolean {
  if (name === null) return true;
  const lowered = name.toLowerCase();
  if (state === "existant") return /exist|actuel/.test(lowered);
  return /projet|project|futur/.test(lowered);
}

function buildParentMap(api: IfcAPI, modelID: number): Map<number, number> {
  const parentMap = new Map<number, number>();
  const relAggIds = api.GetLineIDsWithType(modelID, WebIFC.IFCRELAGGREGATES, true);
  for (let i = 0; i < relAggIds.size(); i++) {
    const rel = api.GetLine(modelID, relAggIds.get(i)) as Record<string, unknown>;
    const parent = attrRef(rel, "RelatingObject");
    if (parent === null) continue;
    for (const child of attrRefList(rel, "RelatedObjects")) parentMap.set(child, parent);
  }
  return parentMap;
}

/** Groups by type name only. Side is *not* decided here: a real IFC export
 * can put both lanes of a road under the same IfcPavementType, or even model
 * both sides as one compound product's mesh (e.g. "all accotements" as one
 * part) — so which physical side a piece of geometry is on can only be
 * determined per width sample (see pavementWidthSamples' offset-gap split in
 * ifcGeometry.ts), not once per product or per type name. */
function listPavementGroups(api: IfcAPI, modelID: number, state: StateKind): PavementGroup[] {
  const pavementIds = api.GetLineIDsWithType(modelID, WebIFC.IFCPAVEMENT, true);
  const ids: number[] = [];
  for (let i = 0; i < pavementIds.size(); i++) ids.push(pavementIds.get(i));

  const parentMap = buildParentMap(api, modelID);
  const roadNames = new Map(ids.map((id) => [id, roadAncestorName(api, modelID, id, parentMap)]));
  const hasDualState = ids.some((id) => {
    const n = roadNames.get(id) ?? null;
    return n !== null && matchesState(n, "existant") !== matchesState(n, "projet");
  });

  const groups = new Map<string, PavementGroup>();
  for (const id of ids) {
    if (hasDualState && !matchesState(roadNames.get(id) ?? null, state)) continue;
    const line = api.GetLine(modelID, id) as Record<string, unknown>;
    const typeName = pavementTypeName(api, modelID, line, id);
    if (!groups.has(typeName)) groups.set(typeName, { typeName, expressIds: [] });
    groups.get(typeName)!.expressIds.push(id);
  }
  return Array.from(groups.values());
}

function slugify(name: string): string {
  return name
    .toLowerCase()
    .replace(/[^a-z0-9]+/g, "_")
    .replace(/^_+|_+$/g, "");
}

export function extractIfcState(
  api: IfcAPI,
  modelID: number,
  state: StateKind,
  axis: AxisReference,
  typeMapping: Map<string, [Side, ElementType]> = new Map(),
): { bands: Band[]; samples: WidthSample[] } {
  const groups = listPavementGroups(api, modelID, state);

  const bands: Band[] = [];
  const samples: WidthSample[] = [];
  for (const group of groups) {
    const allSamples: PlanWidthSample[] = [];
    for (const id of group.expressIds) {
      pushAll(allSamples, pavementWidthSamples(api, modelID, id, axis));
    }

    const bySide = new Map<Side, PlanWidthSample[]>();
    for (const s of allSamples) {
      if (!bySide.has(s.side)) bySide.set(s.side, []);
      bySide.get(s.side)!.push(s);
    }

    for (const [autoSide, sideSamples] of bySide.entries()) {
      const bandId = `ifc-${state}-${slugify(group.typeName)}-${autoSide}`;
      let side: Side = autoSide;
      let elementType: ElementType;
      let source: SourceMethod;
      let confidence: number;
      if (typeMapping.has(bandId)) {
        [side, elementType] = typeMapping.get(bandId)!;
        source = "recuperation_entrees";
        confidence = 1.0;
      } else {
        [elementType, confidence] = guessElementType(group.typeName);
        source = "recuperation_dxf";
      }

      const widths = sideSamples.map((s) => s.width);
      bands.push({
        band_id: bandId,
        state,
        side,
        element_type: elementType,
        source,
        confidence,
        label_hint: group.typeName,
        sample_count: widths.length,
        width_min: widths.length ? minOf(widths) : null,
        width_max: widths.length ? maxOf(widths) : null,
        width_mean: widths.length ? widths.reduce((a, b) => a + b, 0) / widths.length : null,
      });
      for (const s of sideSamples) {
        samples.push({
          pk: s.pk,
          side,
          element_type: elementType,
          state,
          width_m: s.width,
          source,
          band_id: bandId,
          near_x: s.near[0],
          near_y: s.near[1],
          far_x: s.far[0],
          far_y: s.far[1],
        });
      }
    }
  }
  return { bands, samples };
}

File: backend/app/extraction/ifc_extractor.py
"""IfcRoad -> IfcRoadPart -> IfcPavement traversal and width extraction.

Per the brief: IfcPavementType names are user-defined and can be misleading, so
this module only ever *suggests* a (side, element_type) mapping — confidence is
kept low and the caller (API layer) always exposes the suggestion for user
confirmation via the same override mechanism used for DXF bands.
"""

from __future__ import annotations

import re
from dataclasses import dataclass, field

import numpy as np
import ifcopenshell

from app.extraction.axis_reference import AxisReference
from app.extraction.ifc_geometry import pavement_width_samples
from app.models.domain import Band, ElementType, Side, SourceMethod, StateKind, WidthSample

_KEYWORD_HINTS: list[tuple[str, ElementType]] = [
    (r"bau", ElementType.BAU),
    (r"accot", ElementType.ACCOTEMENT),
    (r"tr.{0,4}oir", ElementType.TROTTOIR),
    (r"cycl", ElementType.CYCLE),
    (r"voie|chauss", ElementType.VOIE),
    (r"tpc|median|terre.?plein", ElementType.TPC),
]


def guess_element_type(type_name: str) -> tuple[ElementType, float]:
    """Best-effort keyword guess from the IFC type name. Deliberately low
    confidence: the brief documents a real case where a type named 'Trotoir'
    actually represented the carriageway, so this is a suggestion only."""
    name = type_name.lower()
    for pattern, element_type in _KEYWORD_HINTS:
        if re.search(pattern, name):
            return element_type, 0.3
    return ElementType.NON_UTILISE, 0.1


@dataclass
class PavementGroup:
    type_name: str
    side: Side
    products: list = field(default_factory=list)


def _pavement_type_name(pavement) -> str:
    for rel in getattr(pavement, "IsTypedBy", []) or []:
        rel_type = rel.RelatingType
        if rel_type is not None:
            return rel_type.Name or rel_type.is_a()
    return pavement.Name or f"Pavement-{pavement.id()}"


def _road_ancestor_name(pavement) -> str | None:
    current = pavement
    for _ in range(6):
        rels = getattr(current, "Decomposes", []) or []
        if not rels:
            return None
        parent = rels[0].RelatingObject
        if parent.is_a("IfcRoad"):
            return parent.Name
        current = parent
    return None


def _matches_state(name: str | None, state: StateKind) -> bool:
    if name is None:
        return True
    lowered = name.lower()
    if state == StateKind.EXISTANT:
        return any(k in lowered for k in ("exist", "actuel"))
    return any(k in lowered for k in ("projet", "project", "futur"))


def list_pavement_groups(ifc, state: StateKind, axis: AxisReference) -> list[PavementGroup]:
    """Groups by (type name, côté) rather than type name alone: a real IFC
    export typically gives both lanes of a road the same IfcPavementType
    (e.g. a single "Voie" type used on both sides), so grouping by type name
    only would merge the left and right pavements into one band and force a
    single side label onto their combined geometry -- losing one side
    entirely. Side is guessed per individual pavement product, before
    grouping, so left and right stay distinct bands even when they share a
    type name."""
    pavements = ifc.by_type("IfcPavement")
    road_names = {p.id(): _road_ancestor_name(p) for p in pavements}
    has_dual_state = any(_matches_state(n, StateKind.EXISTANT) != _matches_state(n, StateKind.PROJET) and n for n in road_names.values())

    groups: dict[tuple[str, Side], PavementGroup] = {}
    for p in pavements:
        if has_dual_state and not _matches_state(road_names[p.id()], state):
            continue
        type_name = _pavement_type_name(p)
        side = _guess_side([p], axis)
        key = (type_name, side)
        groups.setdefault(key, PavementGroup(type_name=type_name, side=side)).products.append(p)
    return list(groups.values())


def _guess_side(products, axis: AxisReference) -> Side:
    offsets = []
    for p in products:
        from app.extraction.ifc_geometry import shape_vertices

        verts = shape_vertices(p)
        if verts is None:
            continue
        offsets.extend(axis.axis.project_point((v[0], v[1]))[1] for v in verts[:: max(1, len(verts) // 50)])
    mean_offset = float(np.mean(offsets)) if offsets else 0.0
    return Side.GAUCHE if mean_offset >= 0 else Side.DROITE


def extract_ifc_state(
    path: str,
    state: StateKind,
    axis: AxisReference,
    type_mapping: dict[str, tuple[Side, ElementType]] | None = None,
) -> tuple[list[Band], list[WidthSample]]:
    """`type_mapping` is keyed by band_id (stable: f"ifc-{state}-{slug(type_name)}"),
    matching the same override mechanism used for DXF bands."""
    ifc = ifcopenshell.open(path)
    type_mapping = type_mapping or {}
    groups = list_pavement_groups(ifc, state, axis)

    bands: list[Band] = []
    samples: list[WidthSample] = []
    for group in groups:
        slug = re.sub(r"[^a-z0-9]+", "_", group.type_name.lower()).strip("_")
        band_id = f"ifc-{state.value}-{slug}-{group.side.value}"
        side = group.side
        if band_id in type_mapping:
            side, element_type = type_mapping[band_id]
            source = SourceMethod.RECUPERATION_ENTREES
            confidence = 1.0
        else:
            element_type, confidence = guess_element_type(group.type_name)
            source = SourceMethod.RECUPERATION_DXF

        widths: list[tuple[float, float]] = []
        for product in group.products:
            widths.extend(pavement_width_samples(product, axis))

        band = Band(
            band_id=band_id,
            state=state,
            side=side,
            element_type=element_type,
            source=source,
            confidence=confidence,
            label_hint=group.type_name,
            sample_count=len(widths),
            width_min=min((w for _, w in widths), default=None),
            width_max=max((w for _, w in widths), default=None),
            width_mean=(sum(w for _, w in widths) / len(widths)) if widths else None,
        )
        bands.append(band)
        for pk, width in widths:
            samples.append(
                WidthSample(
                    pk=pk,
                    side=side,
                    element_type=element_type,
                    state=state,
                    width_m=width,
                    source=source,
                    band_id=band_id,
                )
            )
    return bands, samples
